### PR TITLE
fix: UiShape now refreshes when screen size changes

### LINKED ISSUE #814: SDK UI Canvas does not adapt to window resolution
The UIs generated by developers through the SDK does not adapt well to window size changes.

@decentralgames
"When playing, resizing the screen to 50% and then later slowly getting it back to a larger state is causing the bottom UI to get fixed in place instead of flexing."

### PR DESCRIPTION
**Pull request by Kinerius**
## What does this PR change?

Fixes #814 

...

## How to test the changes?

Unity Only:
1. Change branch to `fix/screen-resize-uishape`
2. Start the game with _**Open Browser When Start**_ toggled **OFF**
2. Create a new scene with `dcl init`
3. Add this to game.ts
```typescript
const canvas = new UICanvas()
const board = new UIContainerRect(canvas)
board.name = "board-rect"
board.width = "60%"
board.height = "60%"
board.hAlign = "center"
board.color = Color4.Green()
```

4. Start that scene with `dcl start` then add `&ws=ws%3A%2F%2Flocalhost%3A4999%2Fdcl` to the link
5. Resize the Game view, the UI should adapt to the new screen size

### CODE REVIEW

**Review by Suduck — CHANGES_REQUESTED** [review]
LGTM! However, I think that we need to change the method name or it won't work in production

File: unity-renderer/Assets/Scripts/MainScripts/DCL/Controllers/ScreenController/ScreenBridge.cs [hunk 1]
```diff
@@ -0,0 +1,36 @@
+using System;
+using UnityEngine;
+
+namespace DCL.Components
+{
+    public class ScreenBridge : MonoBehaviour
+    {
+        public static ScreenBridge i { get; private set; }
+        public event Action OnScreenResize;
```
[the remark this answers is not in the document]
Comment (BrianAmadori) [reply unity-renderer/Assets/Scripts/MainScripts/DCL/Controllers/ScreenController/ScreenBridge.cs:9]: I agree. The bridges flow should always try to be outwards. Maybe we can add a new `DataStore` element like "currentResolution" and make the `UIShapes` wait for it to change? We will have the added boon to be able to know the given window resolution.

[the remark this answers is not in the document]
Comment (Kinerius) [reply unity-renderer/Assets/Scripts/MainScripts/DCL/Controllers/ScreenController/ScreenBridge.cs]: Allright, finally decided to go for a simpler KISS approach using a DataStore object to store the screen size. Alas, the MonoBehaviour that watches the screen size is inevitable.
Changes on this [commit](https://github.com/decentraland/unity-renderer/pull/956/commits/4f747820281e70f4d15034067a8f0c47bba1f614)

File: unity-renderer/Assets/Scripts/MainScripts/DCL/Controllers/ScreenController/ScreenBridge.cs [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+        private void LateUpdate()
+        {
+            if (HasScreenSizeChanged())
+            {
+                UpdateScreenSize();
+                SendScreenResizeEvent();
+            }
+        }
+
+        public void ForceScreenResizeEvent() { SendScreenResizeEvent(); }
```
[the remark this answers is not in the document]
Comment (Kinerius) [reply unity-renderer/Assets/Scripts/MainScripts/DCL/Controllers/ScreenController/ScreenBridge.cs]: I removed the kernel messaging for this feature as it had no use for now.

**Review by BrianAmadori — CHANGES_REQUESTED** [review]
We have to add at least a test to ensure this keeps working and we are good to go.

File: unity-renderer/Assets/Scripts/MainScripts/DCL/Components/UI/UIContainer/Tests/UIContainerStackTests.cs [hunk 3]
```diff
@@ -117,6 +118,48 @@ public IEnumerator NormalizedSize()
...
+            // we simulate first that the UI container now has the desired screen size            
+            screenSpaceShape.childHookRectTransform.SetSizeWithCurrentAnchors(RectTransform.Axis.Horizontal, desiredScreenSize.x);
+            screenSpaceShape.childHookRectTransform.SetSizeWithCurrentAnchors(RectTransform.Axis.Vertical, desiredScreenSize.y);
+            
+            //Screen resize event
+            DataStore.i.screen.size.Set(desiredScreenSize);
+            
+            var currentSize = uiContainerStack.childHookRectTransform.rect.size;
+
+            Assert.AreEqual(desiredScreenSize.x * widthPercent/100f, currentSize.x, 0.01f, "UIContainer width after screen resize");
```
Comment (BrianAmadori) [inline unity-renderer/Assets/Scripts/MainScripts/DCL/Components/UI/UIContainer/Tests/UIContainerStackTests.cs:156]: We shouldn't be asserting the size before the size change too just to make sure the event works?

File: unity-renderer/Assets/Scripts/MainScripts/DCL/Components/UI/UIContainer/Tests/UIContainerStackTests.cs [hunk 4]
```diff
@@ -117,6 +118,48 @@ public IEnumerator NormalizedSize()
...
+            Assert.AreEqual(desiredScreenSize.y * heightPercent/100f, currentSize.y, 0.01f, "UIContainer height after screen resize");
+            screenSpaceShape.Dispose();
+            yield return null;
```
Comment (BrianAmadori) [inline unity-renderer/Assets/Scripts/MainScripts/DCL/Components/UI/UIContainer/Tests/UIContainerStackTests.cs:159]: This `yield return null;` is necessary?

**Review by BrianAmadori — APPROVED** [review]
LGTM! Approve with a few minor comments.

File: unity-renderer/Assets/Scripts/MainScripts/DCL/Components/UI/UIContainer/Tests/UIContainerStackTests.cs (hunk 4 above)
Comment (Kinerius) [reply unity-renderer/Assets/Scripts/MainScripts/DCL/Components/UI/UIContainer/Tests/UIContainerStackTests.cs:159]: Its not, that was from a copy-paste from the test above 🥇

File: unity-renderer/Assets/Scripts/MainScripts/DCL/Components/UI/UIContainer/Tests/UIContainerStackTests.cs (hunk 3 above)
Comment (Kinerius) [reply unity-renderer/Assets/Scripts/MainScripts/DCL/Components/UI/UIContainer/Tests/UIContainerStackTests.cs:158]: You are right, fixed it on last commit

**Review by BrianAmadori — APPROVED** [review]

File: unity-renderer/Assets/Scripts/MainScripts/DCL/Components/UI/UIShape.cs [hunk 5]
```diff
@@ -153,7 +151,13 @@ public class Model : BaseModel
 
         public UIShape parentUIComponent { get; protected set; }
 
-        public UIShape() { model = new Model(); }
+        public UIShape()
+        {
+            DataStore.i.screen.size.OnChange += OnScreenResize;
```
Comment (AjimenezDCL) [inline unity-renderer/Assets/Scripts/MainScripts/DCL/Components/UI/UIShape.cs:156]: When working with the `DataStore` I find better to have a property in the class
`
private Vector2Int screenSize => DataStore.i.screen.size;
`

It reduces the amount of "noise" when searching for references when lot of systems are accessing a data.

File: unity-renderer/Assets/Scripts/MainScripts/DCL/Controllers/ScreenController/ScreenSizeWatcher.cs [hunk 6]
```diff
@@ -0,0 +1,31 @@
+using System;
+using UnityEngine;
+
+namespace DCL.Components
+{
+    public class ScreenSizeWatcher : MonoBehaviour
+    {
+        private BaseVariable<Vector2Int> ScreenSize => DataStore.i.screen.size;
```
Comment (AjimenezDCL) [inline unity-renderer/Assets/Scripts/MainScripts/DCL/Controllers/ScreenController/ScreenSizeWatcher.cs:8]:
```suggestion
        private BaseVariable<Vector2Int> screenSize => DataStore.i.screen.size;
```

**Review by AjimenezDCL — APPROVED** [review]
Left a couple of comments.

Looks good! Love the data-driven approach

**Review by Suduck — APPROVED** [review]

**Pull request merged**